In [8]:
import pandas as pd
import numpy as np

In [9]:
class Capa_densa:
    # n_features y n_neuronas nos indica dimendión de la capa
    def __init__(self, n_features, n_neuronas):
        self.weights = 0.01*np.random.rand(n_features, n_neuronas)
        self.bias = np.zeros((1, n_neuronas))

    # Funciones de activación: ReLU, sigmoid
    def relu(self, inputs):
        # outputs = []
        # for i in inputs:
        #     if i > 0:
        #         outputs.append(i)
        #     else:
        #         outputs.append(0)
        return np.maximum(0, inputs)

    def sigmoid(self, inputs): #Clasificación binaria y clasificación multietiqueta
        
        return 1 / (1 + np.exp(np.negative(inputs)))
    
    def softmax(self, inputs): # Clasificación clasificación multiclase etiquetas excluyentes
        
        return np.exp(inputs) / np.exp(inputs).sum(axis = 1, keepdims = True)

    def forward(self, inputs, activation = 'relu'):
        self.output = np.dot(inputs, self.weights) + self.bias
        print(self.output)
        if activation == 'relu':
            self.output = self.relu(self.output)
        if activation == 'sigmoid':
            self.output = self.sigmoid(self.output)
        if activation == 'softmax':
            self.output = self.softmax(self.output)

In [10]:
df = pd.read_csv('Titanic-Dataset.csv')
df = df.drop(columns=['PassengerId', 'Name', 'Ticket', 'Cabin', 'Embarked']).dropna()

df['Sex'] = df['Sex'].map({'male': 0, 'female': 1})
df

Survived  Pclass  Sex   Age  SibSp  Parch     Fare
0           0       3    0  22.0      1      0   7.2500
1           1       1    1  38.0      1      0  71.2833
2           1       3    1  26.0      0      0   7.9250
3           1       1    1  35.0      1      0  53.1000
4           0       3    0  35.0      0      0   8.0500
..        ...     ...  ...   ...    ...    ...      ...
885         0       3    1  39.0      0      5  29.1250
886         0       2    0  27.0      0      0  13.0000
887         1       1    1  19.0      0      0  30.0000
889         1       1    0  26.0      0      0  30.0000
890         0       3    0  32.0      0      0   7.7500

[714 rows x 7 columns]

In [11]:
y_df = np.array(df['Survived'])
x_df = df.drop(columns=['Survived'])

x_df

Pclass  Sex   Age  SibSp  Parch     Fare
0         3    0  22.0      1      0   7.2500
1         1    1  38.0      1      0  71.2833
2         3    1  26.0      0      0   7.9250
3         1    1  35.0      1      0  53.1000
4         3    0  35.0      0      0   8.0500
..      ...  ...   ...    ...    ...      ...
885       3    1  39.0      0      5  29.1250
886       2    0  27.0      0      0  13.0000
887       1    1  19.0      0      0  30.0000
889       1    0  26.0      0      0  30.0000
890       3    0  32.0      0      0   7.7500

[714 rows x 6 columns]

In [12]:
y_df

array([0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1,
       0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1,
       1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1,
       0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1,
       0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1,
       0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1,
       1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1,

### Red neuronal:
- 714 inputs
- 6 features
- 1a capa: 10 neuronas
- Output layer: 2 neuronas (survive vs no) (yo lo hayía con 1 neurona. Cuanto más cercano a 1 más posibilidades de sobrevivir)

In [13]:
# Inicinializamos capas.
capa_1 = Capa_densa(6, 10)
output_layer = Capa_densa(10, 2)

# Pasos forward.

capa_1.forward(x_df)
output_layer.forward(capa_1.output, activation='softmax')

output_layer.output

[[0.24341161 0.16853039 0.18174253 ... 0.14070872 0.10690003 0.19265871]
 [0.8105184  0.83979286 0.60044494 ... 0.61283631 0.64148256 0.40676227]
 [0.27549802 0.19357975 0.20246718 ... 0.15932298 0.1212433  0.22225245]
 ...
 [0.36581306 0.37157508 0.26852699 ... 0.27412949 0.27781706 0.19777244]
 [0.4211629  0.38495224 0.31090769 ... 0.28275437 0.28520668 0.23825476]
 [0.32170486 0.20208478 0.23777627 ... 0.16415084 0.12530595 0.25557005]]
[[0.00866714 0.00889166]
 [0.03380738 0.03153226]
 [0.00989002 0.0101849 ]
 ...
 [0.01516862 0.01426836]
 [0.01651364 0.01582137]
 [0.01094348 0.01142313]]


array([[0.49994387, 0.50005613],
       [0.50056878, 0.49943122],
       [0.49992628, 0.50007372],
       ...,
       [0.50022507, 0.49977493],
       [0.50017307, 0.49982693],
       [0.49988009, 0.50011991]])